File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
# 약관 판단 후 유효기간을 지나 파기되어야 한다면 True 반환
def check(cur_year, cur_month, cur_day, year, month, day, value):
    month += value

    if month > 12:
        year += (month - 1) // 12
        month = (month - 1) % 12 + 1

    if cur_year != year:
        return cur_year > year

    if cur_month != month:
        return cur_month > month

    return cur_day >= day

def solution(today, terms, privacies):
    answer = []
    cur_year, cur_month, cur_day = map(int, today.split('.'))
    expiration = {}

    for term in terms:
        key, value = term.split(' ')
        expiration[key] = int(value)

    for idx, privacie in enumerate(privacies):
        date, key = privacie.split(' ')
        year, month, day = map(int, date.split('.'))
        value = expiration[key]

        # 유효기간 확인
        if check(cur_year, cur_month, cur_day, year, month, day, value):
            answer.append(idx + 1)

    return answer

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
/** 약관 판단 후 유효기간을 지나 파기되어야 한다면 ture 반환 */
function check(curYear, curMonth, curDay, year, month, day, value) {
    month += value
    
    if (month > 12) {
        year += Math.floor((month - 1) / 12);
        month = (month - 1) % 12 + 1;
    }

    
    if (curYear !== year) {
        return curYear > year
    }

    if (curMonth !== month) {
        return curMonth > month
    }

    return curDay >= day
}

function solution(today, terms, privacies) {
    let answer = []
    const [curYear, curMonth, curDay] = today.split('.').map(string => Number(string))
    let expiration = new Map()
    
    terms.map(term => {
        const [key, value] = term.split(' ')
        
        expiration[key] = Number(value)
    })
    
    privacies.map((privacie, idx) => {
        const [date, key] = privacie.split(' ')
        const [year, month, day] = date.split('.').map(string => Number(string))
        const value = expiration[key]
        
        // 유효기간 확인
        if (check(curYear, curMonth, curDay, year, month, day, value)) {
            answer.push(idx + 1)
        }
    })
    
    return answer
}